### PR TITLE
Multi section updates up to Advanced Components

### PR DESCRIPTION
**Pull request by rajinder-yadav**
Changes rolled in from my own comments made on the Angular training Gitbook. Michael suggested to do one PR for all my changes.

### CODE REVIEW

File: handout/advanced-components/access_child_components.md [hunk 1]
```diff
@@ -1,15 +1,14 @@
...
+ ## @ViewChild and @ViewChildren ##
+The @ViewChild and @ViewChildren decorators provide access to the classe of child component from the containing component.
 
-The `@ViewChild` & `@ViewChildren` decorators allow access to the classes of child components of the current component.
+The @ViewChild is a decarator function that takes the name of a component class as its input and finds it's selector in the template of the containing component to bind to. @ViewChild can also be passed a template reference variable.
 
-`@ViewChild` selects one class instance of the given child component class when given a type.
-
-For example, we can call `show` which is on the child component `Alert` class:
+For example, we bind the component class 'Alert' to its selector '<my-alert>' and assign it to the property 'alert'. This allows us to gain access to the functions in that class like Alert.show().
```
Comment (clbond) [inline handout/advanced-components/access_child_components.md]: Should use backticks instead of apostrophes for code snippets so that they get formatted properly

File: handout/advanced-components/access_child_components.md [hunk 2]
```diff
@@ -59,7 +58,7 @@ export class App {
 ```
 [View Example](http://plnkr.co/edit/8Eak9DANedsZDHBHuLea?p=preview)
 
-As shown above, when given a class type `@ViewChild` & `@ViewChildren` select child components based on type. However, they can also be passed selector strings:
+As shown above, given a class type to @ViewChild and @ViewChildren a child component or a list of children component are selected respectively using their selector from the template. In addition both @ViewChild and @ViewChildren can be passed a selector string:
```
Comment (clbond) [inline handout/advanced-components/access_child_components.md]: Should put backticks around `@ViewChild` and `@ViewChildren` since they are code snippets

File: handout/bootstrapping/README.md [hunk 3]
```diff
@@ -1,6 +1,5 @@
 # Bootstrapping an Angular 2 Application
 
-Bootstrapping is an essential process in Angular - it is where the application is built, and where Angular comes to life.
+Bootstrapping is an essential process in Angular - it is where the application is built and loaded, it's where Angular comes to life.
```
Comment (clbond) [inline handout/bootstrapping/README.md]: This statement is misleading in the context of AoT (ahead-of-time compilation).

File: handout/bootstrapping/file_structure.md [hunk 4]
```diff
@@ -56,7 +56,7 @@ platformBrowserDynamic().bootstrapModule(MyAppModule)
 
 [View Example](https://plnkr.co/edit/kVqEKf?p=preview)
 
-This is the main entry point of the application. The `MyAppModule`  operates as the root module of our  application. The module is configured to use `MyApp` as the component to bootstrap, and will be rendered on any `app` HTML element encountered.
+The bootstrap process loads **main.ts** which is the main entry point of the application. The `MyAppModule` operates as the root module of our application. The module is configured to use `MyApp` as the component to bootstrap, and will be rendered on any `app` HTML element encountered.
 
 There is an `app` HTML element in the *index.html* file, and we use *app/index.ts* to import the `MyAppModule` component and the `platformBrowserDynamic().bootstrapModule` function and kickstart the  process.
```
Comment (clbond) [inline handout/bootstrapping/file_structure.md]: Perhaps make a note about optional usage of `bootstrapModuleFactory` for AoT

File: handout/components/app_structure/passing_data_into_components.md [hunk 5]
```diff
@@ -1,6 +1,8 @@
 # Passing Data into a Component
 
-The `inputs` attribute defines a set of parameters that can be passed down from the component's parent. For example, we can modify the `Hello` component so that `name` can be configured by the parent.
+There are two ways to pass data into a component, with 'property binding' and 'event binding'. Data flows top-down from parent into children, while events flow bottom-up from child to parent. Events works just like they do in regular HTML DOM based events and are cancelable.
```
Comment (clbond) [inline handout/components/app_structure/passing_data_into_components.md]: `EventEmitter` emissions are cancellable? As far as I know, this is false. If you were to implement some kind of cancellation policy I think you would have to do it manually. (But maybe there is a way to do it and I just don't know? If so, add more detail to this paragraph on how to cancel an `EventEmitter` event.)

File: handout/components/app_structure/passing_data_into_components.md [hunk 6]
```diff
@@ -1,6 +1,8 @@
...
+
+The `inputs` decarator defines a set of parameters that can be passed down from the component's parent. For example, we can modify the `Hello` component so that `name` can be configured by the parent.
```
Comment (clbond) [inline handout/components/app_structure/passing_data_into_components.md]: `inputs` is not a decorator, it's a property, a la `@Component({inputs: [...]})`, and it is deprecated. This should say `@Input()` instead.

File: handout/components/creating_components.md [hunk 7]
```diff
@@ -1,9 +1,11 @@
 # Creating Components
 
-Components in Angular 2 build upon this idea. We define a component's application logic inside a class. To this we then attach a selector and a template.
+Components in Angular 2 build upon the lessons learned from Angular 1.5. We define a component's application logic inside a class. To this we attach metadata using a TypeScript **Decorator**, which is simply a function call that allows the specification of the **selector** and **template**.
```
Comment (clbond) [inline handout/components/creating_components.md]: Decorator does not need to be capitalized. A decorator is a more generic concept than just allowing `selector` and `template`. This should say something like "a decorator allows you to modify a class or function definition and add metadata to properties and function arguments." Then say something about `selector` and `template`.

File: handout/components/creating_components.md [hunk 8]
```diff
@@ -1,9 +1,11 @@
...
 
-- _selector_ is the element property that we use to tell Angular to create and insert an instance of this component.
-- _template_ is a form of HTML that tells Angular how to render this component.
+- _Selector_ is the element property that we use to tell Angular to create and insert an instance of this component.
```
Comment (clbond) [inline handout/components/creating_components.md]: `selector` and `template` are property names and should not be capitalized (to reduce confusion)

File: handout/components/creating_components.md [hunk 9]
```diff
@@ -1,9 +1,11 @@
...
+- _Template_ is a form of HTML that tells Angular what needs to be to rendered in the DOM.
+
+The Component below will interpolate the value of **name** variable into the template between the double braces '{{name}}', what get rendered in the view is `<p>Hello World</p>`.
```
Comment (clbond) [inline handout/components/creating_components.md]: Backticks not apostrophes

File: handout/components/creating_components.md [hunk 10]
```diff
@@ -21,6 +23,6 @@ export class Hello {
 }
 ```
 
-To use this component we simply add `<hello></hello>` to our HTML, and Angular will insert an instance of the `Hello` view between those tags.
+We need to import the **Component** decarator from '@angular/core' before we can make use of it. To use this component we simply add `<hello></hello>` to the HTML file or another template, and Angular will insert an instance of the `Hello` view between those tags.
```
Comment (clbond) [inline handout/components/creating_components.md]: Decorator, not "decarator". "Component" and "@angular/core" should be in backticks

File: handout/directives/attribute_directives.md [hunk 11]
```diff
@@ -1,5 +1,5 @@
 # Attribute Directives
 
-Attribute directives are a way of changing the appearance or behavior of a component. Ideally, a directive should work in a way that is component agnostic and not bound to implementation details.
+Attribute directives are a way of changing the appearance or behavior of a DOM element. Ideally, a directive should work in a way that is component agnostic and not bound to implementation details.
```
Comment (clbond) [inline handout/directives/attribute_directives.md]: This is misleading. The original wording ("component") made more sense. Attribute directives can be applied to components as well, not just DOM elements.

File: handout/directives/ng_for_directive.md [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # NgFor Directive
 
-The `ngFor` directive is a way of repeating a template by using each item of an iterable as that template's context.
+The `ngFor` directive is a way of repeatly adding content inside the ngFor block to the DOM.
```
Comment (clbond) [inline handout/directives/ng_for_directive.md]: Original wording made more sense

File: handout/directives/structural_directives.md [hunk 13]
```diff
@@ -1,6 +1,6 @@
 # Structural Directives
 
-Structural Directives are a way of handling how a component or element renders through the use of the `template` tag. This allows us to run some code that decides what the final rendered output will be. Angular 2 has a few built-in structural directives such as `ngIf`, `ngFor`, and `ngSwitch`.
+Structural Directives control what is added or removed from the DOM. It allows us to run some code that decides what the final rendered output will be. Angular 2 has a few built-in structural directives such as `ngIf`, `ngFor`, and `ngSwitch`.
```
Comment (clbond) [inline handout/directives/structural_directives.md]: This is misleading. I think the original text made more sense

File: handout/features/README.md [hunk 14]
```diff
@@ -3,7 +3,7 @@
 ![ES6/TypeScript](../images/es6-typescript.png "ES6 and TypeScript")
 
 The language we usually call "JavaScript" is formally known as "EcmaScript".
-The new version of JavaScript, "EcmaScript 6" or "ES6", offers a number of new features that extend the power of the language.
-ES6 is not widely supported in today's browsers, so it needs to be transpiled to ES5. 
-You can choose between several transpilers, but we'll be using TypeScript, 
+The new version of JavaScript, "EcmaScript 2015" or "ES6", offers a number of new features that extend the power of the language.
```
Comment (clbond) [inline handout/features/README.md]: EcmaScript 2015 and ES6 are different standards. Misleading

File: handout/features/README.md [hunk 15]
```diff
@@ -3,7 +3,7 @@
...
+ES6 is not widely supported in today's browsers, so it needs to be transpiled to ES5.
```
Comment (clbond) [inline handout/features/README.md]: ES6 actually is widely supported in modern browsers; it is poorly supported in older browsers. See: http://kangax.github.io/compat-table/es6/

File: handout/features/delegation.md [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  }
+  walk() {
+    this.bird.walk();
+  }
+  swim() {
+    console.log('swim!');
+  }
+}
+
+let bird = new Bird(...);
```
Comment (clbond) [inline handout/features/delegation.md]: Why are we using `let` here instead of `const` when the value never changes?

File: handout/features/es6_modules.md [hunk 17]
```diff
@@ -1,8 +1,33 @@
-# ES6 Modules
+# ECMAScript 2015 Modules
```
Comment (clbond) [inline handout/features/es6_modules.md]: Modules were first defined in ES6 not ES2015, this should be changed back

File: handout/features/es6_modules.md [hunk 18]
```diff
@@ -1,8 +1,33 @@
...
 
-ES6 also introduces the concept of a module, which works similar to other languages. Defining an ES6 module is quite easy: each file is assumed to define a module and we specify its exported values using the `export` keyword. 
 
-Loading ES6 modules is a little trickier. In an ES6-compliant browser you use the `System` keyword to load modules asynchronously. To make our code work with current browsers, however, we will use the SystemJS library as a polyfill:
+ECMAScript 2015 introduced _module_ support. A module in ES2015 is single file that allows code and data to be isolated, it helps in organizing and grouping code logically. In other languages it's called a package or library.
```
Comment (clbond) [inline handout/features/es6_modules.md]: ES6 introduced module support, not ES2015. Misleading

File: handout/features/es6_modules.md [hunk 19]
```diff
@@ -1,8 +1,33 @@
...
+
+All code and data inside the module has file scope, what this means is they are not accessible from code outside the module. To share code or data outside a module, it needs to be exported using the **export** keyword.
+
+```js
+// File: circle.js
+
+export const pi = 3.141592;
+
+export function circumference(diameter) {
```
Comment (clbond) [inline handout/features/es6_modules.md]: Better to use `export const circumference = diameter => diameter * pi` since this section is about ES6

File: handout/features/inheritance.md [hunk 20]
```diff
@@ -1,86 +1,88 @@
...
+## Object Initialization
+
+The class constructor is called when an object is created using the **new** operator, it will be called before the object is fully created. A consturctor is used to pass in arguments to initialize the newly created object.
+
+The order of object creation starts from its base class and then moves down to any subclass(es).
 
-To illustrate this, let's image we have a zoo application where types of birds are created. In the classical inheritance case, we would define an parent class `Bird` first, then let, say, `Penguin` be a descent of `Bird`:
 ```js
-//ES6
+// Base Class (ES2016)
```
Comment (clbond) [inline handout/features/inheritance.md]: ES2016 now? This should say ES6

File: handout/tooling/chrome.md [hunk 21]
```diff
@@ -1,6 +1,6 @@
-## Chrome
+## Web Browser
 
 We use Google's Chrome browser for this course because of its cutting-edge JavaScript engine and excellent debugging tools.
 
-However, code written with AngularJS should work on any modern web browser (Firefox, IE9+, Chrome, Safari).
+Some who champion the freedom of the 'Open Free Internet' use Mozilla Firefox, there is a [Developer Edition](https://www.mozilla.org/en-US/firefox/developer/) available with development and debugging tools. However, code written with JavaScript should work on any modern web browser (Firefox, IE9+, Chrome, Safari, Opera).
```
Comment (clbond) [inline handout/tooling/chrome.md]: Almost everyone in our office uses Chrome, surprised to see a Firefox blurb!

File: handout/advanced-components/access_child_components.md (hunk 1 above)
Comment (rajinder-yadav) [inline handout/advanced-components/access_child_components.md]: Done.

File: handout/advanced-components/access_child_components.md (hunk 2 above)
Comment (rajinder-yadav) [inline handout/advanced-components/access_child_components.md]: Done.

File: handout/components/creating_components.md (hunk 7 above)
Comment (rajinder-yadav) [inline handout/components/creating_components.md]: Nice suggestion, went with yours.

File: handout/components/creating_components.md (hunk 8 above)
Comment (rajinder-yadav) [inline handout/components/creating_components.md]: Good point, corrected.

File: handout/components/creating_components.md (hunk 9 above)
Comment (rajinder-yadav) [inline handout/components/creating_components.md]: Corrected.

File: handout/directives/attribute_directives.md (hunk 11 above)
Comment (rajinder-yadav) [inline handout/directives/attribute_directives.md]: From the Angular site, it uses DOM elements, "An Attribute directive changes the appearance or behavior of a DOM element."

This change makes the correction.

File: handout/directives/ng_for_directive.md (hunk 12 above)
Comment (rajinder-yadav) [inline handout/directives/ng_for_directive.md]: We'll it doesn't repeat the entire template, this is incorrect, only the content inside the ngFor is repeated. If you have HTML and other angular "template" code outside ngFor score, that will not be repeated. Here I believe there is an error in the Angular 2 doc which use templates and not content. I think my change makes more sense, we ask an additional reviewer to comment?

File: handout/directives/structural_directives.md (hunk 13 above)
Comment (rajinder-yadav) [inline handout/directives/structural_directives.md]: Take from Angular doc, under "Template Syntax" section, it say this, "When NgIf is false, Angular physically removes the element subtree from the DOM".

The changes follow the ng2 doc.

File: handout/features/README.md (hunk 14 above)
Comment (rajinder-yadav) [inline handout/features/README.md]: Actually ECMAScript 2015 is the same as ES6, the committed made the decision to use years. We should follow the recommendation. Here is something wikipedia:

https://en.wikipedia.org/wiki/ECMAScript#6th_Edition_-_ECMAScript_2015

Changed Ecma to ECMA.

File: handout/features/es6_modules.md (hunk 17 above)
Comment (rajinder-yadav) [inline handout/features/es6_modules.md]: As pre previous comment both are the same, going with year.

File: handout/features/es6_modules.md (hunk 18 above)
Comment (rajinder-yadav) [inline handout/features/es6_modules.md]: Will correct 'ES2015' to ECMAScript 2015' are per previous comment.

File: handout/features/es6_modules.md (hunk 19 above)
Comment (rajinder-yadav) [inline handout/features/es6_modules.md]: I was using noob JS! Will make the update.

File: handout/features/inheritance.md (hunk 20 above)
Comment (rajinder-yadav) [inline handout/features/inheritance.md]: Typo, should be 2015.

File: handout/tooling/chrome.md (hunk 21 above)
Comment (rajinder-yadav) [inline handout/tooling/chrome.md]: Ah ya, this could turn into a political hot topic, was trying to keep rangle.io un-biased. Students see us as leaders and will just go with what we say, suggest. Also FF lives and dies off it's install base, just keeping things open and fair.

**Comment by barretodavid** [comment]
@rajinder-yadav this is a massive PR, it's very hard to find the time to do a complete review. In the future please consider doing smaller PRs, maybe focused on one section at a time.

File: handout/features/interfaces.md [hunk 22]
```diff
@@ -1,18 +1,10 @@
 # Interfaces
 
-Sometimes classes are "more" than a programmer wants. Classes end up creating
-code, in the form of transpiled ES6 classes or transpiled ES5 constructor
-functions.
+An _interface_ is a TypeScript artifact, it is not part of ECMAScript. A _interface_ is a way to define a _contract_ on a function with respect to the arguments and their type. Along with functions, an _interface_ can also be used with a Class as well define user defined types.
```
Comment (barretodavid) [inline handout/features/interfaces.md]: Change:

> can also be used with a Class as well define user defined types.

To:

> can also be used with a Class as well **to define custom** types.

File: handout/features/interfaces.md [hunk 23]
```diff
@@ -1,18 +1,10 @@
...
 
-Also, JavaScript is a _subset_ of TypeScript, and in JavaScript functions are
-"first class" (they can be assigned to variables and passed around), so how can
-functions be described in TypeScript?
+An interface is an abstract type, it does not contain any code as a _class_ does. It's only defines the 'signature' or shape of an API. During the translation stage a TypeScript interface will not generate any code, it only used only by TypeScript for type checking.
```
Comment (barretodavid) [inline handout/features/interfaces.md]: > **It** only defines the signature...

Comment (barretodavid) [inline handout/features/interfaces.md]: Proposed change:

> During transpilation, an `interface` will not generate any code, it is only used by Typescript for type checking during development.

File: handout/features/template_strings.md [hunk 24]
```diff
@@ -14,7 +14,7 @@ console.log('hello my name is ' + name + ' I am ' + age + ' years old');
 ```
 
 ES6 introduces a new type of string literal that is marked with back ticks (`).
-These string literals _can_ include newlines, and there is a new mechanism for
+These string literals _can_ include newlines, and there is a string interpolation for
 inserting variables into strings:
```
Comment (barretodavid) [inline handout/features/template_strings.md]: different line?

File: handout/tooling/module_loading_bundling_and_build_tasks.md [hunk 25]
```diff
@@ -1,4 +1,4 @@
 # Module Loading, Bundling and Build Tasks: [Webpack](http://webpack.github.io/docs/what-is-webpack.html)
 
-Webpack takes modules with dependencies and generates static assets representing those modules. It can bundle JavaScript, CSS, HTML or just about anything via additional loaders. Webpack can also be extended via plugins, for example minification and mangling can be done using the UglifyJS plugin for webpack.
+Webpack is a JavaScript module bundler. It takes modules with their dependencies and generates static assets representing those modules. Webpack known only how to  bundle JavaScript. To bundle other assets likes CSS, HTML, images or just about anything it uses additional loaders and plugins. Webpack can also be extended via plugins, for example minification and mangling can be done using the UglifyJS plugin for webpack.
```
Comment (barretodavid) [inline handout/tooling/module_loading_bundling_and_build_tasks.md]: I think it is correct to only mention loaders in this sentence (not plugins):

> To bundle other assets likes CSS, HTML, images or just about anything it uses additional loaders and plugins

File: handout/directives/ng_for_directive.md (hunk 12 above)
Comment (winkerVSbecks) [inline handout/directives/ng_for_directive.md]: Original wording LGTM too. You can wrap content inside ngFor or use a template tag instead. In either case nothing gets added inside the ngFor. You just end up with multiple copies of the thing you attached the ngFor to.

File: handout/directives/attribute_directives.md (hunk 11 above)
Comment (rajinder-yadav) [inline handout/directives/attribute_directives.md]: Updated with new suggestion.

File: handout/directives/ng_for_directive.md (hunk 12 above)
Comment (rajinder-yadav) [inline handout/directives/ng_for_directive.md]: Reverted to original. Now make sense what Chris was alluding to, forgot out the syntax sugar!

File: handout/directives/structural_directives.md (hunk 13 above)
Comment (rajinder-yadav) [inline handout/directives/structural_directives.md]: Reverted!

File: handout/features/es6_modules.md (hunk 17 above)
Comment (rajinder-yadav) [inline handout/features/es6_modules.md]: Eventually the long format is going to be the only way going forward, so we might as well adhere to the standard now. I feel we should just get rid of ES6 altogether and use the long format, early on we can mention ES6 == ECMAScript 2015. Thoughts?

File: handout/advanced-components/access_child_components.md [hunk 26]
```diff
@@ -1,15 +1,14 @@
  # Accessing Child Component Classes #
 
- ## @ViewChild & @ViewChildren ##
+ ## @ViewChild and @ViewChildren ##
+The @ViewChild and @ViewChildren decorators provide access to the classe of child component from the containing component.
 
-The `@ViewChild` & `@ViewChildren` decorators allow access to the classes of child components of the current component.
+The `@ViewChild` is a decarator function that takes the name of a component class as its input and finds it's selector in the template of the containing component to bind to. @ViewChild can also be passed a template reference variable.
```
Comment (rajinder-yadav) [inline handout/advanced-components/access_child_components.md]: Corrected.

File: handout/advanced-components/access_child_components.md [hunk 27]
```diff
@@ -1,15 +1,14 @@
...
 
-`@ViewChild` selects one class instance of the given child component class when given a type.
-
-For example, we can call `show` which is on the child component `Alert` class:
+For example, we bind the component class `Alert` to its selector `<my-alert>` and assign it to the property `alert`. This allows us to gain access to the functions in that class like Alert.show().
```
Comment (rajinder-yadav) [inline handout/advanced-components/access_child_components.md]: Good point, changes incorporated.

File: handout/advanced-components/access_child_components.md [hunk 28]
```diff
@@ -30,10 +29,10 @@ export class App {
 
 In the interest of separation of concern, we'd normally want to have child elements take care of their own behaviors and pass in an `@Input()`. However, it might be a useful construct in keeping things generic.
 
-We can also use `@ViewChildren` to get a list of class instances if there are multiple, which selects a `QueryList` of the elements:
+When there are multiple embedded components in the template, we make use of @ViewChildren to collect an instance of its component class inside a QueryList container. QueryList is a TypeScript generic class that simply states it is a list container of class Alert. The embedded component in the template is identified by it's selector, in the code below that would be each '<my-alert>' selector.
```
Comment (rajinder-yadav) [inline handout/advanced-components/access_child_components.md]: Updated.

File: handout/features/es6_modules.md (hunk 17 above)
Comment (barretodavid) [inline handout/features/es6_modules.md]: I agree that we should mention the difference between ES6, ES2015 and ECMAScript 2015 at the beginning. I wouldn't use ECMAScript 2015 everywhere because is just too long but we could use the shorthand ES2015 which people are going to get used to when working with tools like Webpack or Typescript.

File: handout/advanced-components/access_child_components.md (hunk 28 above)
Comment (rajinder-yadav) [inline handout/advanced-components/access_child_components.md]: Updated with suggested paragraph.

File: handout/components/app_structure/passing_data_into_components.md [hunk 29]
```diff
@@ -1,6 +1,8 @@
 # Passing Data into a Component
 
-The `inputs` attribute defines a set of parameters that can be passed down from the component's parent. For example, we can modify the `Hello` component so that `name` can be configured by the parent.
+There are two ways to pass data into a component, with 'property binding' and 'event binding'. Data flows top-down from parent into children, while events flow bottom-up from child to parent. Events works just like they do in regular HTML DOM based events, which makes event propagation cancelable.
```
Comment (rajinder-yadav) [inline handout/components/app_structure/passing_data_into_components.md]: You're right, with Angular 2 flow is always top down on each event loop, both for ng2 events and ng2 data flow. Updated with rewording!

File: handout/components/app_structure/passing_data_into_components.md [hunk 30]
```diff
@@ -1,6 +1,8 @@
...
+
+The `@Input()` decarator defines a set of parameters that can be passed down from the component's parent. For example, we can modify the `Hello` component so that `name` can be configured by the parent.
```
Comment (rajinder-yadav) [inline handout/components/app_structure/passing_data_into_components.md]: Changed.

File: handout/features/README.md [hunk 31]
```diff
@@ -3,7 +3,7 @@
 ![ES6/TypeScript](../images/es6-typescript.png "ES6 and TypeScript")
 
 The language we usually call "JavaScript" is formally known as "EcmaScript".
-The new version of JavaScript, "EcmaScript 6" or "ES6", offers a number of new features that extend the power of the language.
-ES6 is not widely supported in today's browsers, so it needs to be transpiled to ES5. 
-You can choose between several transpilers, but we'll be using TypeScript, 
-which is what the Angular team uses to write Angular 2. Angular 2 makes use of a number of features of ES6 and TypeScript.
+The new version of JavaScript, "ECMAScript 2015" previously known as "ES6", offers a number of new features that extend the power of the language.
```
Comment (rajinder-yadav) [inline handout/features/README.md]: As we spoke making global update to use ES6 for now, mention ES2015 being the same thing for now.

File: handout/features/classes.md [hunk 32]
```diff
@@ -13,6 +13,24 @@ class Hamburger {
...
 ```
 
-Traditional class-based languages often reserve the word `this` to reference the
-current (runtime) instance of the class.  This is also true in JavaScript, but
-JavaScript code can _optionally_ supply `this` to a method at call time.
+Traditional class-based languages often reserve the word `this` to reference the current (runtime) instance of the class. In Javascript `this` refers to the calling context and therefore can change to be something other than the object.
+
+## Object
+
+An object is an instance of a class which is created using operator `new`. When using a dot notation to access a method on the object, `this` will refer to the object to the left of the dot.
```
Comment (rajinder-yadav) [inline handout/features/classes.md]: updated.

File: handout/features/classes.md [hunk 33]
```diff
@@ -13,6 +13,24 @@ class Hamburger {
...
+
+```js
+let burger = new Hamburger();
+burger.listToppings();
+```
+
+Above whenever `this` is used from inside class Hamburger, it will refer to object `burger`.
```
Comment (rajinder-yadav) [inline handout/features/classes.md]: Update with suggestion.

File: handout/features/classes.md [hunk 34]
```diff
@@ -13,6 +13,24 @@ class Hamburger {
...
+
+## Changing Caller Context
+JavaScript code can _optionally_ supply `this` to a method at call time using one of the following.
```
Comment (rajinder-yadav) [inline handout/features/classes.md]: Added.

File: handout/features/classes.md [hunk 35]
```diff
@@ -13,6 +13,24 @@ class Hamburger {
...
+
+* Function.prototype.call(object [,arg, ...])
+* Function.prototype.bind(object [,arg, ...])
+* Function.prototype.apply(object [,argsArray])
```
Comment (rajinder-yadav) [inline handout/features/classes.md]: Let's break this into a separate PR, will need to consider this more.

File: handout/features/delegation.md [hunk 36]
```diff
@@ -0,0 +1,37 @@
+# Delegation
+
+In the inheritance section we looked at one way to extend a class functionality, there is second way using delegation to extend functionality. With delegation, one object will contain a reference to a different object that it will hand off a request to perform the functionality.
+
+The code below shows how to use delegation with the Bird class and Penguin class. The Penguin class how a reference to the Bird class and it delegrates call made to it's _walk_ method over to Bird's _walk_ method.
```
Comment (rajinder-yadav) [inline handout/features/delegation.md]: Backticks added, 2nd part typo changed 'how' to 'has'.

File: handout/features/es6_modules.md [hunk 37]
```diff
@@ -16,3 +47,4 @@ Loading ES6 modules is a little trickier. In an ES6-compliant browser you use th
       });
   </script>
 ```
+
```
Comment (rajinder-yadav) [inline handout/features/es6_modules.md]: The joy of JS, new standards every month :D ... I'll the suggestion.

File: handout/features/inheritance.md [hunk 38]
```diff
@@ -1,86 +1,88 @@
 ## Inheritance
 
-JavaScript's inheritance works differently from inheritance in other languages, which can be very confusing. ES6 classes provide a syntactic sugar attempting to alleviate the issues with using prototypical inheritance present in ES5. Our recommendation is still to avoid using inheritance or at least deep inheritance hierarchies. Try solving the same problems through delegation instead.
+JavaScript's inheritance works differently from inheritance in other languages, which can be very confusing. ES6 classes provide a syntactic sugar attempting to alleviate the issues with using prototypical inheritance present in ES5.
+
+To illustrate this, let's image we have a zoo application where types of birds are created. In classical inheritance, we define a base class and then subclass it to create a derived class.
+
+## Subclassing
+
+The example code below shows how to derive Penguin from Bird using the **extend** keyword. Also pay attention to the **super** keyword used in the subclass constructor of Penguin, it is used to pass the argument to the base class Bird's constructor.
```
Comment (rajinder-yadav) [inline handout/features/inheritance.md]: corrected.

File: handout/features/inheritance.md [hunk 39]
```diff
@@ -1,86 +1,88 @@
...
+
+The Bird class defines the method _walk_ which is inherited by the Penguin class and is available for use by instance of Penguin objects. Likewise the Penguin class defines the method _swim_ which is not avilable to Bird objects. Inheritance works top-down from base class to its subclass.
+
+## Object Initialization
+
+The class constructor is called when an object is created using the **new** operator, it will be called before the object is fully created. A consturctor is used to pass in arguments to initialize the newly created object.
```
Comment (rajinder-yadav) [inline handout/features/inheritance.md]: Fixed.

File: handout/features/interfaces.md (hunk 22 above)
Comment (rajinder-yadav) [inline handout/features/interfaces.md]: Going with suggestion, however UDT (user defined define) is widely used.

File: handout/features/interfaces.md (hunk 23 above)
Comment (rajinder-yadav) [inline handout/features/interfaces.md]: Corrected and updated with suggestion.

File: handout/features/template_strings.md (hunk 24 above)
Comment (rajinder-yadav) [inline handout/features/template_strings.md]: There were other, all broken lines were joined.

File: handout/tooling/module_loading_bundling_and_build_tasks.md (hunk 25 above)
Comment (rajinder-yadav) [inline handout/tooling/module_loading_bundling_and_build_tasks.md]: Corrected, removed plugins.

File: handout/features/classes.md (hunk 35 above)
Comment (barretodavid) [inline handout/features/classes.md]: Good idea

File: handout/features/es6_modules.md (hunk 37 above)
Comment (barretodavid) [inline handout/features/es6_modules.md]: The note as originally there, look at the diff:

> Loading ES6 modules is a little trickier. In an ES6-compliant browser you use the `System` keyword to load modules asynchronously. To make our code work with current browsers, however, we will use the SystemJS library as a polyfill:

**Comment by barretodavid** [comment]
@rajinder-yadav you're almost there. There's only two comments remaining: Bootstraping + AoT and `System.import` as polyfill

File: handout/bootstrapping/file_structure.md (hunk 4 above)
Comment (rajinder-yadav) [inline handout/bootstrapping/file_structure.md]: AoT text added.

File: handout/features/es6_modules.md (hunk 37 above)
Comment (rajinder-yadav) [inline handout/features/es6_modules.md]: Updated with suggestion, thanks David.

**Comment by barretodavid** [comment]
Good job @rajinder-yadav, your PR is ready to be merged now
